cvat/apps/engine/views.py

```diff
@@ -696,20 +696,28 @@ def __call__(self, request, start, stop, db_data):
                 return HttpResponse(buf.getvalue(), content_type=mime)
 
             elif self.type == 'context_image':
-                if not (start <= self.number <= stop):
-                    raise ValidationError('The frame number should be in ' +
-                        f'[{start}, {stop}] range')
-
-                image = Image.objects.get(data_id=db_data.id, frame=self.number)
-                for i in image.related_files.all():
-                    path = os.path.realpath(str(i.path))
-                    image = cv2.imread(path)
-                    success, result = cv2.imencode('.JPEG', image)
-                    if not success:
-                        raise Exception('Failed to encode image to ".jpeg" format')
-                    return HttpResponse(io.BytesIO(result.tobytes()), content_type='image/jpeg')
-                return Response(data='No context image related to the frame',
-                    status=status.HTTP_404_NOT_FOUND)
+                if start <= self.number <= stop:
+                    zip_buffer = io.BytesIO()
+                    image = Image.objects.get(data_id=db_data.id, frame=self.number)
+                    with zipfile.ZipFile(zip_buffer, 'a', zipfile.ZIP_DEFLATED, False) as zip_file:
+                        if not image.related_files.count():
+                            return Response(data='No context image related to the frame',
+                                status=status.HTTP_404_NOT_FOUND)
+
+                        common_path = os.path.commonpath(list(map(lambda x: str(x.path), image.related_files.all())))
+                        for i in image.related_files.all():
+                            path = os.path.realpath(str(i.path))
+                            name = os.path.relpath(str(i.path), common_path)
+                            image = cv2.imread(path)
+                            success, result = cv2.imencode('.JPEG', image)
+                            if not success:
+                                raise Exception('Failed to encode image to ".jpeg" format')
+                            zip_file.writestr(f'{name}.jpg', result.tobytes())
+                            # response = HttpResponse(wrapper, content_type='application/zip')
+                            # response['Content-Disposition'] = 'attachment; filename=your_zipfile.zip'
+                    return HttpResponse(io.BytesIO(zip_buffer.getvalue()), content_type='application/zip')
```

Comment:
Probably HttpStreamingResponse is better. Let's have HttpResponse here because in other places we also use it. Another idea is to use FileStream (I believe it is possible to use io.BytesIO together with the response type).